text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+
+This expression is allowed in `const` contexts, but is not promotable.
+
+*Note: the behaviour of this expression is independent of whether the type gets
+layout-optimized.  For example, the following holds even if `x` is `2_i8` in memory.*
+```rust
+enum MyOption<T> { MyNone, MySome(T) }
+let x = MyOption::<std::cmp::Ordering>::MyNone;
+assert_eq!(x.enum#discriminant, 0_isize);
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This isn't strictly necessary, we could continue to get along just fine without it.
+
+- For the FFI cases the layout guarantees mean it's already possible to write a
+  sound and reliable function that reads the discriminant.
+- For cases without `repr(int)`, custom discriminants aren't even allowed,
+  so those discriminants much not be all that important.
+- It's always possible to write a `match` in safe code that optimizes away
+  and produces exactly the same thing that this new expression would.
+- A pseudo-field with `#` in the name looks kinda weird.
+- There might be a nicer way to do this in the future.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Why have a `#` in the name?
+
+By not being an identifier, `.enum#discriminant` can't conflict with anything.
+
+While today there are no fields directly accessible from values of enum type,
+there are lots of plausible-enough proposals that would allow some.
+
+For example, *enum variant types* have come up repeatedly, which would represent a single
+variant and thus would allow accessing the fields on that type, but plausibly would
+still offer access to the discriminant.  Similarly, a *pattern type* that restricts
+the enum to a single variant would plausibly allow access to its fields.  And one
+of those fields might be named `discriminant`.
+
+Other requests have come in too, like allowing field access if every variant has
+a field with the same name & type or allowing field access if there's only a
+single inhabited variant.
+
+By being clearly different it means it can't conflict with any field or method.
+That also helps resolve any concerns about it *looking* like field access -- as
+existed for `.await` -- since it's visibly lexically different.
+
+And the lexical space is already reserved,
+
+## Why have `enum` in the name?
+
+Well, it seemed short and evocative enough to be fine.
+Doing something like `e#` isn't shorter enough to matter, and
+I'd rather save very-short prefixes for higher-prevalence things.
+
+And since it's a pre-existing keyword, it means that
+```rust
+let d = foo().bar.enum#discriminant;
+```
+already gets highlighting on the `enum` in my editor without needing any updates.
+
+## Isn't this kinda long?
+
+Not really, compared to the existing possibilities.
+
+For example, in a macro expansion even the internal magic today ends up being
+```rust
+let __self_tag = ::core::intrinsics::discriminant_value(self);
+let __arg1_tag = ::core::intrinsics::discriminant_value(other);
```

Comment:
This should probably not say "tag" when it means "discriminant". The two are not the same in general.

Reply:
this is just a copy of what is being expanded from today's `#[derive(PartialOrd)]` :shrug: it was renamed from `__self_vi` (meaning Index Value) to `__self_tag` by rust-lang/rust#99046.

Reply:
Good point. https://github.com/rust-lang/rust/pull/123919 should fix that.

The RFC should still be updated. :)